CRM/Mailing/BAO/TrackableURL.php

```diff
@@ -72,11 +72,11 @@ public static function getTrackerURL($url, $mailing_id, $queue_id) {
       }
       $id = $tracker->id;
 
-      $redirect = CRM_Utils_System::externUrl('extern/url', "u=$id");
+      $redirect = CRM_Utils_System::previouslyExternUrl('civicrm/mailing/url', ['u=' => $id]);
       $urlCache[$mailing_id . $url] = $redirect;
     }
 
-    $returnUrl = CRM_Utils_System::externUrl('extern/url', "u=$id&qid=$queue_id");
+    $returnUrl = CRM_Utils_System::previouslyExternUrl('civicrm/mailing/url', ['u=' => $id, 'qid' => $queue_id]);
```

Comment:
I think it's great that this patch is providing a way to process the click-through via normal routing/controller (ie `CRM_Mailing_Page_Url`).

But the API for requesting the URL is... getting weird...

The point of introducing `CRM_Utils_System::externUrl()` was to de-couple the *usage* of the URL from the *generation* the URL. As long as consumers go through `CRM_Utils_System::externUrl()`, we can evolve the actual content of the URL.

I don't want to belabor the point, but perhaps it merits an example:

* Historically, `flexmailer:OpenTracker.php` composes links to `extern/open.php`, as in:
    * `$config->userFrameworkResourceURL . "extern/open.php?q=" . $task->getEventQueueId()`
* We want to be able to change that URL. But we don't know what it's going to be in the future, and the substance will depend on the version#s of other things (`civicrm-core`, `civicrm-wp-rest`, etc).  Instead, `flexmailer:OpenTracker.php` should call a function -- that function will *figure out* the appropriate URL. To wit:
    * `CRM_Utils_System::externUrl('extern/open', ...)`
* So now `flexmailer` is up-to-spec. We can mix/match different versions/modules/URLs, and they'll work together. Huzza!
* But when it comes time to actually change the URL, the API is changed again... to `previouslyExternUrl()`. Now `flexmailer:OpenTracker.php` is using incorrect URLs because it's still calling `externUrl()`.

I feel like a policy change in "what is the exact default URL" should be encapsulated s.t. `externUrl()` still works.

Reply:
I've started working on an update for this PR...